### PR TITLE
Microverse - Enumerable Methods

### PR DESCRIPTION
**Pull request by Maclenn77**
I'm adding **Enumerable Methods** to Develop branch.

## List of Methods

- **My_each** Require a block and applied in each item of an array
- **my_each_with_index** Require a block and applied in each item of an array and its index
- **my_select** Require a block and return items selected by the criteria given by the block_given
- **my_all?** Check each item of an Enumerable and return true if all items are true, else false, depending on the criteria given by a block
- **my_any?** Check each item of an Enumerable and return True, if any item is True, else false, depending on the criteria given by a block
- **my_none?** Check each item of an Enumerable and return True, if none item is true, else false, depending on the criteria given by a block.
- **my_count** Return total of items with the criteria given by a block or argument. If any block or argument is added, it will return the number of items of an Enumerable.
- **my_map, my_map_proc, my_map_proc_yield** Returns a new array following the criteria of a block_given
- **my_inject** Apply a process to two items, keep the result and applied to the next one.
- **mupliply_els** Multiply the items of a given Enumerable between them

**Thanks for reviewing!**

### CODE REVIEW

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Maclenn77      
#### Great Job so far, but there are a few more things required. 
### General
- [x] `my_all?`  
- returns true if the block never returns false or nil
  `  block =  proc { |num| num < 200 } `
    `  p [2,32,32].my_all?(&block)  `should return `true`
- when no block or argument is given returns true when none of the collection members are false or nil:
 `p [1, true, 'hi', []].my_all?` should return `true`.
- when a class is passed as an argument returns true if all of the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return `true`
-  when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['d','d'].my_all?(/d/)` should return `true`
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`p [3,3,3,3].my_all?(3)` should return `true`


- [x] `my_none?`:
- returns true if the block never returns true for all elements
` block = proc { |num| num < 2 }`
` p [2,23,21,12].my_none?(&block)` should return `true`
- When no block or argument is given returns true only if none of the collection members is true
`p  [nil, false, true, []].my_none?`  should return `false`
- When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`p ['hello'].my_none?(5)` - should return `true`
- when a class is passed as an argument returns true if none of the collection is a member of such class
`p [3,3,3,3].my_none?(String)`  should return `true`
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
`p ['hello','milk','milo'].my_none?(/z/)` should return `true`

- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`
- Combines all elements of enum by applying a binary operation, specified by a block:
 `p (1..5).my_inject(4) { |prod, n| prod * n }` should return 480


- [x] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(Integer)` should return `true`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('cat')` should return `true`




### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by Maclenn77** [comment]
Thanks for the review, @kobiyoyo !

### Changes in the branch:

- My methods work for Range class
- Fixed problems with my_all? my_any? my_none? and my_inject

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Maclenn77     
#### Great Job so far, You are almost there my friend just few more things required. 
### General

- [x] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
- `p range.my_inject(2, :*)` should return `40320`
- Combines all elements of enum by applying a binary operation, specified by a block:
- `p (3..8).my_inject(4) { |prod, n| prod * n }` should return `80640`
`





### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by Maclenn77** [comment]
Thanks, @kobiyoyo ! I didn't notice that problem with my_inject

### Changes in the branch:

- Fixed problems with my_inject

File: My_Methods.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  # My Map
+  def my_map
+    return enum_for unless block_given?
+
+    arr = []
+    my_each do |i|
+      arr << yield(i)
+    end
+    arr
+  end
```
Comment (euqueme) [inline My_Methods.rb:98]: [OPTIONAL]
- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

**Review by euqueme — APPROVED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Maclenn77 :handshake: 

### Thank you for doing the changes requested by the previous reviewer, your project is very professional, your Readme is well explained and you have used good ruby conventions, your enumerable methods are also working properly so you may move on! :tada: :tada: :tada: 

### Please :pray: merge this branch and keep up the good work :clinking_glasses: 

PROJECT STATUS: APPROVED :green_apple:

File: My_Methods.rb (hunk 1 above)
Comment (Maclenn77) [reply My_Methods.rb:98]: Hi, @euqueme !

I spent a lot of time trying to figure out how to take either a proc or a block. I googled it, I asked my mentor, and I even [open a question in StackOverflow](https://stackoverflow.com/questions/61237942/is-there-any-way-to-handling-an-exception-for-rubys-syntaxerror-both-arg-and-a).

If you can give me a hint to achieve that, I would be sincerily thankful.

**Pull request merged**